### ISSUE TITLE
Campaign Timeline doesn't load for large campaigns

### ISSUE THREAD

**Issue by W1R3D-Code** [gophish/gophish#765]
Get a Highcharts error in console when loading campaign page with more events than turboThreshold value (defaults to 1000)

`Highcharts error #12: www.highcharts.com/errors/12`

> Highcharts Error #12
> 
> Highcharts expects point configuration to be numbers or arrays in turbo mode
> 
> This error occurs if the series.data option contains object configurations and the number of points exceeds the turboThreshold. It can be fixed by either setting the turboThreshold option to a higher value, or changing your point configurations to numbers or arrays. See turboThreshold.

**Comment by jordan-wright** [comment]
Hmmm..... this one is interesting. I'll have to troubleshoot this and get back to you.

Thanks for the heads up!

**Comment by jordan-wright** [comment]
I've added a temporary fix by disabling the `turboThreshold`. Longer term, though, this will require a bit more thinking.

Here's some notes (mainly so I can come back and look through them later):

Highcharts has a turbo feature which lets it assume certain attributes about the data points. This works fine for [x,y] arrays, but not for objects where things are more complex. In our case, we store the email and the event in the point so we can render the right tooltip. So, when there's more events than the `turboThreshold` allows, Highcharts thinks it'll be too expensive to render the chart. In the short term, I've tested with a campaign with 5k events and haven't seen too bad of performance.

In the long term, I'd like to leverage their [boost](https://www.highcharts.com/blog/news/175-highcharts-performance-boost/) module which does clever HTML5 canvas -> svg conversions to make rendering large amounts of points a breeze. But, this has the same problem - we need our points as x,y pairs, not objects.

So I'll have to figure out a way to store just the x,y pairs and reference an external array. This would need to stay updated when polling for new events, etc. That's future-Jordan's problem 😉
